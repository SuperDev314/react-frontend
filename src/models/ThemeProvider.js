import React, { PropTypes, Component } from 'react'
import { isFunction, isPlainObject } from 'lodash'
import createBroadcast from '../utils/create-broadcast'

export const CHANNEL = '__styled-components__'

/**
 * Provide a theme to an entire react component tree via context and event listeners (have to do
 * both context and event emitter as pure components block context updates)
 */
class ThemeProvider extends Component {
  constructor() {
    super()
    this.getTheme = this.getTheme.bind(this)
    this.getBroadcastsContext = this.getBroadcastsContext.bind(this)
  }

  componentWillMount() {
    // If there is a ThemeProvider wrapper anywhere around this theme provider, merge this theme
    // with the outer theme
    if (this.context.broadcasts && this.context.broadcasts[CHANNEL]) {
      const subscribe = this.context.broadcasts[CHANNEL]
      this.unsubscribeToOuter = subscribe(theme => {
        this.outerTheme = theme
      })
    }
    this.broadcast = createBroadcast(this.getTheme())
  }

  getChildContext() {
    return {
      broadcasts: this.getBroadcastsContext(),
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.theme !== nextProps.theme) this.broadcast.publish(this.getTheme(nextProps.theme))
  }

  componentWillUnmount() {
    this.unsubscribeToOuter()
  }

  // Merge new broadcast with existing ones
  getBroadcastsContext() {
    return Object.assign({}, this.context.broadcasts, { [CHANNEL]: this.broadcast.subscribe })
  }

  // Get the theme from the props, supporting both (outerTheme) => {} as well as object notation
  getTheme(passedTheme) {
    const theme = passedTheme || this.props.theme
    if (isFunction(theme)) {
      const mergedTheme = theme(this.outerTheme)
      if (!isPlainObject(mergedTheme)) {
        throw new Error('[ThemeProvider] Please return an object from your theme function, i.e. theme={() => ({})}!')
      }
      return mergedTheme
    }
    if (!isPlainObject(theme)) {
      throw new Error('[ThemeProvider] Please make your theme prop a plain object')
    }
    return Object.assign({}, this.outerTheme, theme)
  }

  render() {
    if (!this.props.children) {
      return null
    }
    return React.Children.only(this.props.children)
  }
}

ThemeProvider.propTypes = {
  children: PropTypes.node,
  theme: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.object,
  ]),
}
ThemeProvider.childContextTypes = {
  broadcasts: PropTypes.object.isRequired,
}
ThemeProvider.contextTypes = {
  broadcasts: PropTypes.object,
}

export default ThemeProvider
