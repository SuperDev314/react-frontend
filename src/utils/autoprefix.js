import camelizeStyleName from 'fbjs/lib/camelizeStyleName'
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName'
import { autoprefix } from 'glamor/lib/autoprefix'

export default root => {
  root.walkDecls(decl => {
    const objStyle = { [camelizeStyleName(decl.prop)]: decl.value }
    const prefixed = autoprefix(objStyle)
    Object.keys(prefixed).reverse().forEach(newProp => {
      decl.cloneBefore({
        prop: hyphenateStyleName(newProp),
        value: prefixed[newProp],
      })
    })
    decl.remove()
  })
}
