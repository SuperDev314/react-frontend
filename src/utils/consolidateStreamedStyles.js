// @flow
/**
 * When using streaming rendering, style blocks are emitted in chunks directly.
 * This method used to be necessary to move all style elements before the
 * rehydration kicks in. However the rehydration now takes all SSR style tags,
 * combines them into a single tag, and adds it to the head.
 * This renders the consolidation useless
 */
export default function consolidateStreamedStyles() {
  /* noop */
}
