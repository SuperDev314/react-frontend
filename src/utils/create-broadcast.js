/**
 * Creates a broadcast that can be listened to, i.e. simple event emitter
 *
 * @see https://github.com/ReactTraining/react-broadcast
 */
const createBroadcast = (initialValue) => {
  let listeners = []
  let currentValue = initialValue

  return {
    publish(value) {
      currentValue = value
      listeners.forEach(listener => listener(currentValue))
    },
    subscribe(listener) {
      listeners.push(listener)

      // Publish to this subscriber once immediately.
      listener(currentValue)

      // eslint-disable-next-line no-return-assign
      return () =>
        listeners = listeners.filter(item => item !== listener)
    },
  }
}

export default createBroadcast
