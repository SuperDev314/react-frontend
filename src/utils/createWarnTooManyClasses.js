// @flow

import styleSheet from '../models/StyleSheet'

const LIMIT = 200

export default () => {
  let generatedClasses = {}
  let warningSeen = false

  return (className: string, displayName: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true
      if (Object.keys(generatedClasses).length >= LIMIT) {
        const latestRule = styleSheet.rules().find(s => s.selectorText && s.selectorText.indexOf(`.${className}`) !== -1)
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(`Over ${LIMIT} classes were generated for component ${displayName}. ` +
           'Consider using style property for frequently changed styles.\n' +
           'Example: const MyStyledComponent = styled(props => <div className={props.className} style={{color: props.color}}>{props.children}</div>)`width: 100px; height: 100px;`' +
           (latestRule ? `\nLatest generated class: ${latestRule.cssText}` : ''))
        warningSeen = true
        generatedClasses = {}
      }
    }
  }
}
