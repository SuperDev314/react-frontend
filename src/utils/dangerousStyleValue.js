// @flow
import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function dangerousStyleValue(name: string, value: any): any {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  // $FlowFixMe
  if (typeof value === 'symbol') {
    return value;
  }

  const isEmpty = value == null || typeof value === 'boolean' || value === '';
  if (isEmpty) {
    return '';
  }

  if (
    typeof value === 'number' &&
    value !== 0 &&
    !(unitless.hasOwnProperty(name) && unitless[name] === 1)
  ) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
