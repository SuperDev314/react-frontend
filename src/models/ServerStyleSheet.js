// @flow
/* eslint-disable no-underscore-dangle */
import React from 'react'
import stream from 'stream'

import { IS_BROWSER, SC_STREAM_ATTR } from '../constants'
import StyleSheet from './StyleSheet'
import StyleSheetManager from './StyleSheetManager'

declare var __SERVER__: boolean

/* this error is used for makeStyleTag */
const sheetClosedErr =
  process.env.NODE_ENV !== 'production'
    ? `
Can't collect styles once you've consumed a ServerStyleSheet's styles!
ServerStyleSheet is a one off instance for each server-side render cycle.
- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?
`.trim()
    : ''

const streamBrowserErr =
  process.env.NODE_ENV !== 'production'
    ? 'Streaming SSR is only supported in a Node.js environment; Please do not try to call this method in the browser.'
    : ''

export default class ServerStyleSheet {
  closed: boolean
  global: StyleSheet
  instance: StyleSheet

  constructor() {
    this.instance = StyleSheet.global.clone()
    this.closed = false
  }

  collectStyles(children: any) {
    if (this.closed) {
      throw new Error(sheetClosedErr)
    }

    return (
      <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>
    )
  }

  getStyleTags(): string {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toHTML()
  }

  getStyleElement() {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toReactElements()
  }

  interleaveWithNodeStream(readableStream: stream.Readable) {
    if (!__SERVER__ || IS_BROWSER) {
      throw new Error(streamBrowserErr)
    }

    /* the tag index keeps track of which tags have already been emitted */
    const { instance } = this
    let instanceTagIndex = 0

    const streamAttr = `${SC_STREAM_ATTR}="true"`
    const ourStream = new stream.Readable()
    // $FlowFixMe
    ourStream._read = () => {}

    readableStream.on('data', chunk => {
      const { tags } = instance
      let html = ''

      /* retrieve html for each new style tag */
      for (; instanceTagIndex < tags.length; instanceTagIndex += 1) {
        const tag = tags[instanceTagIndex]
        html += tag.toHTML(streamAttr)
      }

      /* force our StyleSheets to emit entirely new tags */
      instance.sealAllTags()
      /* prepend style html to chunk */
      ourStream.push(html + chunk)
    })

    readableStream.on('end', () => {
      this.closed = true
      ourStream.push(null)
    })

    readableStream.on('error', err => {
      this.closed = true
      ourStream.emit('error', err)
    })

    return ourStream
  }
}
