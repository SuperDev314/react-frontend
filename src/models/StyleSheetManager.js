// @flow
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import StyleSheet, { CONTEXT_KEY } from './StyleSheet'
import ServerStyleSheet from './ServerStyleSheet'
import { tagConstructorWithTarget } from './BrowserStyleSheet'

class StyleSheetManager extends Component {
  sheetInstance: StyleSheet
  props: {
    sheet?: StyleSheet | null,
    target?: HTMLElement | null,
  }

  getChildContext() {
    return { [CONTEXT_KEY]: this.sheetInstance }
  }

  componentWillMount() {
    if (this.props.sheet) {
      this.sheetInstance = this.props.sheet
    } else if (this.props.target) {
      this.sheetInstance = new StyleSheet(
        tagConstructorWithTarget(this.props.target)
      )
    } else {
      throw new Error('StyleSheetManager expects either a sheet or target prop')
    }
  }

  render() {
    /* eslint-disable react/prop-types */
    // Flow v0.43.1 will report an error accessing the `children` property,
    // but v0.47.0 will not. It is necessary to use a type cast instead of
    // a "fixme" comment to satisfy both Flow versions.
    return React.Children.only((this.props: any).children)
  }
}

StyleSheetManager.childContextTypes = {
  [CONTEXT_KEY]: PropTypes.oneOfType([
    PropTypes.instanceOf(StyleSheet),
    PropTypes.instanceOf(ServerStyleSheet),
  ]).isRequired,
}

StyleSheetManager.propTypes = {
  sheet: PropTypes.oneOfType([
    PropTypes.instanceOf(StyleSheet),
    PropTypes.instanceOf(ServerStyleSheet),
  ]),
  target: PropTypes.shape({
    appendChild: PropTypes.func.isRequired,
  }),
}

export default StyleSheetManager
