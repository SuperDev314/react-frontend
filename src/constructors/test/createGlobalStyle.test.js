// @flow
import React from 'react';
import ReactDOM from 'react-dom';
import ReactDOMServer from 'react-dom/server';

import { expectCSSMatches, getCSS, resetStyled, resetCreateGlobalStyle, stripComments, stripWhitespace } from '../../test/utils'
import ThemeProvider from '../../models/ThemeProvider'
import ServerStyleSheet from '../../models/ServerStyleSheet'
import StyleSheetManager from '../../models/StyleSheetManager'

const createGlobalStyle = resetCreateGlobalStyle()
const styled = resetStyled();

let context;

beforeEach(() => {
  context = setup()
})

afterEach(() => {
  context.cleanup()
})

describe(`createGlobalStyle`, () => {
  it(`returns a function`, () => {
    const Component = createGlobalStyle``
    expect(typeof Component).toBe('function')
  });

  it(`injects global <style> when rendered`, () => {
    const { render } = context
    const Component = createGlobalStyle`[data-test-inject]{color:red;} `
    render(<Component />)
    expectCSSMatches(`[data-test-inject]{color:red;} `)
  });

  it(`injects global <style> when rendered to string`, () => {
    const sheet = new ServerStyleSheet();
    const Component = createGlobalStyle`[data-test-inject]{color:red;} `
    const html = context.renderToString(sheet.collectStyles(<Component />))

    const container = document.createElement('div');
    container.innerHTML = sheet.getStyleTags();
    const style = container.querySelector('style');

    expect(html).toBe('');
    expect(stripWhitespace(stripComments(style.textContent))).toBe('[data-test-inject]{ color:red; } ');
  });

  it(`supports interpolation`, () => {
    const { cleanup, render } = setup()
    const Component = createGlobalStyle`div {color:${props => props.color};} `
    render(
      <Component color="orange" />
    )
    expectCSSMatches(`div{color:orange;} `)
    cleanup()
  })

  it(`supports theming`, () => {
    const { cleanup, render } = setup()
    const Component = createGlobalStyle`div {color:${props => props.theme.color};} `
    render(
      <ThemeProvider theme={{ color: 'black' }}>
        <Component />
      </ThemeProvider>
    )
    expectCSSMatches(`div{color:black;} `)
    cleanup()
  })

  it(`updates theme correctly`, () => {
    const { cleanup, render } = setup()
    const Component = createGlobalStyle`div {color:${props => props.theme.color};} `
    let update;
    class App extends React.Component {
      state = { color: 'grey' }

      constructor() {
        super()
        update = (payload) => {
          this.setState(payload)
        }
      }

      render() {
        return (
          <ThemeProvider theme={{ color: this.state.color }}>
            <Component />
          </ThemeProvider>
        );
      }
    }
    render(<App />)
    expectCSSMatches(`div{color:grey;} `)

    update({ color: 'red' })
    expectCSSMatches(`div{color:red;} `)

    cleanup()
  })

  it(`renders to StyleSheetManager.target`, () => {
    const { container, render } = context
    const Component = createGlobalStyle`[data-test-target]{color:red;} `
    render(
      <StyleSheetManager target={container}>
        <Component />
      </StyleSheetManager>
    )

    const style = container.firstChild;
    expect(style.tagName).toBe('STYLE')
    expect(style.textContent).toContain(`[data-test-target]{color:red;}`)
  });

  it(`adds new global rules non-destructively`, () => {
    const { container, render } = context
    const Color = createGlobalStyle`[data-test-add]{color:red;} `
    const Background = createGlobalStyle`[data-test-add]{background:yellow;} `

    render(
      <React.Fragment>
        <Color />
        <Background />
      </React.Fragment>
    )

    setTimeout(() => {
      expectCSSMatches(`
        [data-test-add]{color:red;}
        [data-test-add]{background:yellow;}
      `)
    })
  })

  it(`stringifies multiple rules correctly`, () => {
    const { cleanup, render } = setup()
    const Component = createGlobalStyle`
      div {
        color: ${props => props.fg};
        background: ${props => props.bg};
      }
    `
    render(
      <Component fg="red" bg="green" />
    )
    expectCSSMatches(`div{color:red;background:green;} `)
    cleanup()
  })

  it(`injects multiple <GlobalStyle> components correctly`, () => {
    const { cleanup, render } = setup()

    const A = createGlobalStyle`body { background: palevioletred; }`;
    const B = createGlobalStyle`body { color: white; }`;

    render(
      <React.Fragment>
        <A />
        <B />
      </React.Fragment>
    )
    expectCSSMatches(`body{background:palevioletred;} body{color:white;}`)
    cleanup()
  })

  it(`removes styling injected styling when unmounted`, () => {
    const { cleanup, container, render } = setup()
    const Component = createGlobalStyle`[data-test-remove]{color:grey;} `

    class Comp extends React.Component {
      state = {
        styled: true
      }

      componentDidMount() {
        this.setState({ styled: false })
      }

      render() {
        return this.state.styled ? <Component /> : null
      }
    }

    render(<Comp />)
    expect(getCSS(document)).not.toContain(`[data-test-remove]{color:grey;}`)
    cleanup()
  })

  it(`removes styling injected for multiple <GlobalStyle> components correctly`, () => {
    const { container, render } = context

    const A = createGlobalStyle`body { background: palevioletred; }`;
    const B = createGlobalStyle`body { color: white; }`;

    class Comp extends React.Component {
      state = {
        a: true,
        b: true
      }

      onClick() {
        if (this.state.a === true && this.state.b === true) {
          this.setState({
            a: true,
            b: false
          })
        } else if (this.state.a === true && this.state.b === false) {
          this.setState({
            a: false,
            b: false
          })
        } else {
          this.setState({
            a: true,
            b: true
          })
        }
      }

      render() {
        return (
          <div data-test-el onClick={() => this.onClick()}>
            {this.state.a ? <A /> : null}
            {this.state.b ? <B /> : null}
          </div>
        )
      }
    }

    render(<Comp />)
    const el = document.querySelector('[data-test-el]')
    expectCSSMatches(`body{background:palevioletred;} body{color:white;}`)

    {
      el.dispatchEvent(new MouseEvent('click', { bubbles: true }))
      const css = getCSS(document)
      expect(css).not.toContain('body{color:white;}')
      expect(css).toContain('body{background:palevioletred;}')
    }

    {
      el.dispatchEvent(new MouseEvent('click', { bubbles: true }))
      const css = getCSS(document)
      expect(css).not.toContain('body{color:white;}')
      expect(css).not.toContain('body{background:palevioletred;}')
    }
  })

  it(`should throw error when children are passed as props`, () => {
    const { cleanup, render } = setup()
    const Component = createGlobalStyle`
      div {
        color: ${props => props.fg};
        background: ${props => props.bg};
      }
    `
    expect(() => render(
      <Component fg="red" bg="green">
        <div />
      </Component>
    )).toThrowErrorMatchingSnapshot()

    cleanup()
  })
})

function setup() {
  const container = document.createElement('div')
  document.body.appendChild(container)

  return {
    container,
    render(comp) {
      ReactDOM.render(comp, container)
    },
    renderToString(comp) {
      return ReactDOMServer.renderToString(comp)
    },
    cleanup() {
      resetStyled()
      resetCreateGlobalStyle()
      document.body.removeChild(container)
    }
  }
}
