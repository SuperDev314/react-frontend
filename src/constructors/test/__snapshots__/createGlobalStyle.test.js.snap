// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGlobalStyle should throw error when children are passed as props 1`] = `"[createGlobalStyle] received children which will not be rendered. Please use the component without passing children elements."`;
